EO 1097 refactor rxform to forms

khaziwallis opened the pull request:
JIRA: https://jira.rax.io/browse/EO-1097

### LGTMs
- [x] Dev LGTM
- [x] Dev+Vidyo LGTM
- [x] Design LGTM
- [ ] QE LGTM


Before
---------
Components > rxform
![screen shot 2016-07-29 at 9 20 55 pm](https://cloud.githubusercontent.com/assets/3723469/17254434/7a752c9c-55d2-11e6-89d5-fa2dd595273b.png)



After
---------
Elements > Forms
![screen shot 2016-08-01 at 8 32 24 pm](https://cloud.githubusercontent.com/assets/3723469/17298351/384cf6b8-5827-11e6-8813-724b63ee9877.png)
![screen shot 2016-08-01 at 8 32 39 pm](https://cloud.githubusercontent.com/assets/3723469/17298350/38466d70-5827-11e6-899b-113d9e343e91.png)
![screen shot 2016-08-01 at 8 32 48 pm](https://cloud.githubusercontent.com/assets/3723469/17298348/38444be4-5827-11e6-8bf1-ce4888b53819.png)
![screen shot 2016-08-01 at 8 32 58 pm](https://cloud.githubusercontent.com/assets/3723469/17298349/38458c0c-5827-11e6-9051-6823b6b0d857.png)


Simple text controls
---------
![screen shot 2016-08-12 at 7 08 44 pm](https://cloud.githubusercontent.com/assets/3723469/17623953/48b9ab44-60c0-11e6-9593-bef1d6cedb59.png)


text area control
---------
![screen shot 2016-08-12 at 7 08 51 pm](https://cloud.githubusercontent.com/assets/3723469/17623944/3dd726d4-60c0-11e6-8084-eb10d07bb254.png)




Intermediate Controls
---------
![screen shot 2016-08-12 at 7 09 36 pm](https://cloud.githubusercontent.com/assets/3723469/17623971/58400f22-60c0-11e6-8041-ff8190008003.png)


Drop-Down Selection
---------
![screen shot 2016-08-12 at 7 09 56 pm](https://cloud.githubusercontent.com/assets/3723469/17623978/63beaf0c-60c0-11e6-9b32-cff4d7b876e2.png)


Input Groups
---------
![screen shot 2016-08-12 at 7 10 12 pm](https://cloud.githubusercontent.com/assets/3723469/17623994/70b5c2d6-60c0-11e6-907f-bbc79406ecfe.png)


OptionTable
---------
![screen shot 2016-08-11 at 4 46 48 pm](https://cloud.githubusercontent.com/assets/3723469/17590144/2554db94-5ff6-11e6-9974-32ed3d0c543b.png)
![screen shot 2016-08-11 at 4 46 53 pm](https://cloud.githubusercontent.com/assets/3723469/17590146/25597d48-5ff6-11e6-8aff-09bad1d14b15.png)
![screen shot 2016-08-11 at 4 47 00 pm](https://cloud.githubusercontent.com/assets/3723469/17590145/2554c7b2-5ff6-11e6-9a25-ae264f0a90d4.png)




Advanced Controls
---------
![screen shot 2016-08-12 at 7 10 52 pm](https://cloud.githubusercontent.com/assets/3723469/17624019/87bea8d0-60c0-11e6-91f0-56518709bc28.png)




Advanced Text Area
---------
![screen shot 2016-08-17 at 4 36 41 pm](https://cloud.githubusercontent.com/assets/3723469/17737481/ecb30ed2-64aa-11e6-8bbb-54c6a04779ae.png)



Form Actions
---------
![screen shot 2016-08-17 at 4 36 29 pm](https://cloud.githubusercontent.com/assets/3723469/17737469/dfee5d32-64aa-11e6-9dab-0cc9962036c5.png)

Droogans commented:
@khaziwallis it appears as though there has a been a significant visual regression...it looks like the table of data that used to appear under "Layout / Page Example / Layout 3: Form Page" is now missing. This includes a javascript exception that is thrown on the page as well.

```
Error: [ng:areq] Argument 'rxFormDemoCtrl' is not a function, got undefined
```

Visit http://localhost:9001/#/layout/page/form in this branch to see what I'm talking about.

khaziwallis commented:
@glynnis  @CITguy  please take a look for review.

khaziwallis commented:
@Droogans have addressed and made test running with success. please take a look.

CITguy commented:
@khaziwallis, please rebase

CITguy commented:
`grunt server` is reporting issues when compiling ngDocs

![screen shot 2016-08-01 at 11 12 24 am](https://cloud.githubusercontent.com/assets/545605/17300636/df4fb5b2-57d8-11e6-8c53-e1210eb76f64.png)

CITguy commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,101 @@
+<form rx-form name="advancedControlsForm" ng-controller="advancedControlsDemoCtrl">
```
Please update:

* `name="formAdvancedControls"`
* `ng-controller="formAdvancedControlsDemoCtrl"`

CITguy commented on src/elements/Forms/docs/examples/forms.advancedcontrols.js:
```diff
@@ -0,0 +1,7 @@
+angular.module('demoApp')
+.controller('advancedControlsDemoCtrl', function ($scope) {
```
Please rename to `formAdvancedControlsDemoCtrl`

CITguy commented on src/elements/Forms/docs/examples/forms.advancedtextarea.html:
```diff
@@ -0,0 +1,19 @@
+<form rx-form name="advancedTextAreaForm">
```
Please change `name` to `formAdvancedTextArea`

CITguy commented on src/elements/Forms/docs/examples/forms.dropdown.html:
```diff
@@ -0,0 +1,32 @@
+<form rx-form name="dropDownForm" ng-controller="dropDownDemoCtrl">
```
Please update 

* change to `ng-controller="formDropDownDemoCtrl"`
* remove `name="dropDownForm"` (the `name` attribute is unnecessary unless validation is taking place)

CITguy commented on src/elements/Forms/docs/examples/forms.dropdown.js:
```diff
@@ -0,0 +1,86 @@
+angular.module('demoApp')
+.controller('dropDownDemoCtrl', function ($scope) {
```
Please update to `formDropDownDemoCtrl`

CITguy commented on src/elements/Forms/docs/examples/forms.formactions.html:
```diff
@@ -0,0 +1,8 @@
+<form rx-form name="formActionsForm">
```
Please remove `name="formActionsForm"` (the `name` attribute is unnecessary unless validation is taking place).

CITguy commented on src/elements/Forms/docs/examples/forms.inputgroups.html:
```diff
@@ -0,0 +1,108 @@
+<form rx-form name="inputGroupsForm" ng-controller="inputGroupsDemoCtrl">
```
Please update:

* change to `ng-controller="formInputGroupsDemoCtrl"`
* remove `name="inputGroupsForm"` (the `name` attribute is unnecessary unless validation is taking place)

CITguy commented on src/elements/Forms/docs/examples/forms.inputgroups.js:
```diff
@@ -0,0 +1,86 @@
+angular.module('demoApp')
+.controller('inputGroupsDemoCtrl', function ($scope) {
```
Please update to `formInputGroupsDemoCtrl`

CITguy commented on src/elements/Forms/docs/examples/forms.intermediatecontrols.html:
```diff
@@ -0,0 +1,48 @@
+<form rx-form ng-controller="intermediateControlsDemoCtrl" name="intermediateControlsForm">
```
Please update:

* change to `ng-controller="formIntermediateControlsDemoCtrl"`
* change to `name="formIntermediateControls"`

CITguy commented on src/elements/Forms/docs/examples/forms.intermediatecontrols.js:
```diff
@@ -0,0 +1,36 @@
+angular.module('demoApp')
+.controller('intermediateControlsDemoCtrl', function ($scope) {
```
Please update to `formIntermediateControlsDemoCtrl`

CITguy commented on src/elements/Forms/docs/examples/forms.optiontable.html:
```diff
@@ -0,0 +1,65 @@
+<form rx-form name="optionTableForm" ng-controller="optionTableDemoCtrl">
```
Please update:

* change to `ng-controller="formOptionTableDemoCtrl"`
* remove `name="optionTableForm"` (the `name` attribute is unnecessary unless validation is taking place)

CITguy commented on src/elements/Forms/docs/examples/forms.optiontable.js:
```diff
@@ -1,5 +1,5 @@
 angular.module('demoApp')
-.controller('rxFormDemoCtrl', function ($scope) {
+.controller('optionTableDemoCtrl', function ($scope) {
```
Please update to `formOptionTableDemoCtrl`

CITguy commented on src/elements/Forms/docs/examples/forms.simple.html:
```diff
@@ -0,0 +1,36 @@
+<form rx-form name="simpleForm">
```
Please remove `name="simpleForm"` (the `name` attribute is unnecessary unless validation is taking place).

CITguy commented on src/elements/Forms/docs/examples/forms.textarea.html:
```diff
@@ -0,0 +1,16 @@
+<form rx-form name="textAreaForm">
```
Please remove `name="textAreaForm"`

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+    </rx-help-text>
+  </li>
+  <li>
+    <h3>With <code>controlled-width</code> section attribute</h3>
```
`<h3>` tags are unnecessary and semantically invalid. Please remove `<h3>` tags.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
 
-<h3><rx-permalink>Form Validation</rx-permalink></h3>
+<!-- start: rxForm -->
+<h2><rx-permalink>Simple rxForm controls</rx-permalink></h2>
```
Please update:

* This should be an `<h3>` heading.
* Please update text to `Simple Controls`

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<!-- start: rxForm -->
+<h2><rx-permalink>Simple rxForm controls</rx-permalink></h2>
+
+<h3><rx-permalink>Text Inputs</rx-permalink></h3>
```
This should be an `<h4>` heading. Please update.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+</p>
+<rx-example name="forms.simple"></rx-example>
+
+<h3><rx-permalink>Text Area</rx-permalink></h3>
```
This should be an `<h4>` heading. Please update.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+</p>
+<rx-example name="forms.intermediatecontrols"></rx-example>
+
+<h3><rx-permalink>Drop-Down Selection</rx-permalink></h3>
```
This should be an `<h4>` heading. Please update.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+</ul>
+<rx-example name="forms.inputgroups"></rx-example>
+
+<h3><rx-permalink>Option table</rx-permalink></h3>
```
Please update to `Option Tables`

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+</ul>
+<rx-example name="forms.advancedcontrols"></rx-example>
+
+<h3><rx-permalink>Advanced Text Area</rx-permalink></h3>
```
This should be an `<h4>` heading. Please update.

CITguy commented:
@khaziwallis, my first round of feedback has been added. Please update.

khaziwallis commented:
@Droogans yes i can see first tab data is missing. so any referencing PR.. to solve this.

glynnis commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+  <li>
+    <div>
+      <rx-help-text>
+        <p>Oh look! It's using <code>rx-character-count</code></p>
```
This text should be part of the demo itself, rather than part of the documentation text. Please move within the demo and set up the same way as rxForm had this demo functioning, with the lorem ipsum text beside the character count box.

![screen shot 2016-08-02 at 9 55 58 am](https://cloud.githubusercontent.com/assets/1529366/17333298/76db9e50-5897-11e6-9a1e-e6dab02edba1.png)

glynnis replied:
For the documentation text here in the `<p>`, you can put "Text areas can utilize multiple features simultaneously. In the following example, the text area is a required field, has a character counter applied, and shows help text."

khaziwallis replied:
@CITguy  when i try to use these comments inside the broken forms.example file. its giving unexpected border. suggestion how to use comments inside forms.textarea.html file

glynnis replied:
It looks like the field description is appearing under rather than next to the text area with your revisions:
![screen shot 2016-08-03 at 9 50 31 am](https://cloud.githubusercontent.com/assets/1529366/17369902/bfea3d54-595f-11e6-9c5e-7f2518934384.png)

It also appears that the `<p>` text I suggested above hasn't been added above the demo for this area.

Please update to make these changes.

glynnis commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<rx-example name="forms.advancedcontrols"></rx-example>
+
+<h4><rx-permalink>Advanced Text Area</rx-permalink></h4>
+<ul class="list">
```
This class is not necessary and can be removed, along with the following `<li>` and `<div>`.

glynnis commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,101 @@
+  <rx-form-section stacked>
+    <!-- Radio Option + Select -->
+    <rx-field>
+      <rx-field-name>Radio + Select</rx-field-name>
```
I think this should be an `<h4>` to match other demos (see https://github.com/rackerlabs/encore-ui/pull/1762/files#diff-b9acec165521455231e757de340896ccR38)

glynnis commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,101 @@
+
+    <!-- Checkbox + Text -->
+    <rx-field>
+      <rx-field-name>Checkbox + Text Input</rx-field-name>
```
I think this should be an `<h4>` to match other demos (see https://github.com/rackerlabs/encore-ui/pull/1762/files#diff-b9acec165521455231e757de340896ccR38)

CITguy replied:
@khaziwallis: please address this comment.

khaziwallis replied:
@CITguy  does it require to spilt into
Radio + Select
checkbox + Text Input
Dropdown + Text Input + Dropdown
Search Box

CITguy replied:
No split required for this one.

CITguy replied:
This line should be changed from `<rx-field-name>` to `<h5>`, though.

glynnis commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<rx-example name="forms.advancedtextarea"></rx-example>
+
+<h3><rx-permalink>Form Actions</rx-permalink></h3>
+<ul class="list">
```
The `ul` and `li` here are unnecessary. 

Instead, please add this text under the `h3`:
```
<p>
  At the end of each form, the user should be presented with a cancel button and a submit button. The 
  submit button should appear on the left, the cancel button to the right. Submit button text should be as 
  specific as possible to reinforce for the user what change they are about to make. For example, "Update 
  Server" is better than a general "Submit".
</p>
```

glynnis replied:
Please add in the `<p>` text from this comment under the header for this section.

glynnis commented on src/elements/Forms/docs/examples/forms.optiontable.html:
```diff
@@ -0,0 +1,65 @@
+
+  <rx-form-section>
+    <rx-field>
+      <rx-field-name>Empty Option Table</rx-field-name>
```
I think this should be an <h4> to match other demos (see https://github.com/rackerlabs/encore-ui/pull/1762/files#diff-b9acec165521455231e757de340896ccR38)

CITguy replied:
@khaziwallis: please address this comment.

glynnis commented on src/elements/Forms/docs/examples/forms.optiontable.html:
```diff
@@ -0,0 +1,65 @@
+
+  <rx-form-section>
+    <rx-field>
+      <rx-field-name>Checkbox Option Table</rx-field-name>
```
I think this should be an `<h4>` to match other demos (see https://github.com/rackerlabs/encore-ui/pull/1762/files#diff-b9acec165521455231e757de340896ccR38)

CITguy replied:
@khaziwallis: please address this comment.

glynnis commented on src/elements/Forms/docs/examples/forms.optiontable.html:
```diff
@@ -0,0 +1,65 @@
+<form rx-form ng-controller="formOptionTableDemoCtrl">
+  <rx-form-section>
+    <rx-field>
+      <rx-field-name>Radio Option Table</rx-field-name>
```
I think this should be an `<h4>` to match other demos (see https://github.com/rackerlabs/encore-ui/pull/1762/files#diff-b9acec165521455231e757de340896ccR38)

CITguy replied:
This example looks fairly lengthy. I would recommend splitting this into multiple examples.

1. Radio Option Table
2. Checkbox Option Table
3. Empty Option Table

CITguy replied:
@khaziwallis: please address this comment.

glynnis commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
 
+<!-- end: rxForm -->
 
 <h2><rx-permalink>Saving Form Data</rx-permalink></h2>
```
The AutoSave example appears to be broken. The input field doesn't have the proper styles and the field also doesn't save any text if you navigate away from the page and come back.

![screen shot 2016-08-02 at 10 32 49 am](https://cloud.githubusercontent.com/assets/1529366/17334627/9019a0b0-589c-11e6-91a2-1a4d63f84c98.png)

glynnis replied:
This still needs to be addressed.

CITguy replied:
This will be addressed in FRMW-1011. This is a styling issue not introduced by this PR and is thus out of scope.

glynnis commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
-    Some components will style based on both classes being present (typical
-    use case), while others will only use <code>.ng-invalid</code>.
+    Multiple Selection
   </li>
```
Please remove the `ul` and `li`s here -- they're unnecessary. Please add this text in a `<p>` under the `<h4>` instead:
```
For drop-down selections, two options are available: a drop-down that only permits a single selection, or a drop-down that allows the user to make multiple selections.
```

glynnis commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+  <li>
+    Toggle Switches
+  </li>
+</ul>
```
Lines 723-733 can be removed.

glynnis commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+  <li>
+    Empty Option Table
+  </li>
+</ul>
```
Lines 737-747 can be removed.

glynnis commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+      The fields will still stack if horizontal space is constrained, but they will not exceed a reasonable field width.
+    </rx-help-text>
+  </li>
+</ul>
```
Please remove `ul` and `li`s here. Note that the text in lines 753-762 should appear as part of the demo, as it did initially on the rxForm page:
![screen shot 2016-08-02 at 10 48 17 am](https://cloud.githubusercontent.com/assets/1529366/17335198/c127330a-589e-11e6-9a84-1f2154b3f675.png)

glynnis commented:
@khaziwallis Once comments are addressed, I'm happy to take another look at this for design review 👍

khaziwallis commented:
@glynnis @CITguy  have addressed review comments. please take a look.

glynnis commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,55 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<h4><rx-permalink>Drop-Down Selection</rx-permalink></h4>
+<p>
+  For drop-down selections, two options are available: a drop-down that only permits a single selection, or a drop-down that allows the user to make multiple selections.
+  Add a line note
```
Please remove line 716

CITguy commented:
There's a console error at `Layout > Page Examples > Layout 3: Form Page` that is related to updates in this PR. This is also breaking in visual regression PRs.

<img width="871" alt="screen shot 2016-08-03 at 1 38 14 pm" src="https://cloud.githubusercontent.com/assets/545605/17377646/36eb0b2a-5980-11e6-851a-19690e782a0e.png">

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,54 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<h3><rx-permalink>Input Groups</rx-permalink></h3>
+<rx-example name="forms.inputgroups"></rx-example>
+
+<h3><rx-permalink>Option tables</rx-permalink></h3>
```
Heading should have proper capitalization. Please change to `Option Tables`.

CITguy replied:
Heading should be an `<h4>` to fall under "Intermediate Controls". Please update.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,54 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+</p>
+<rx-example name="forms.dropdown"></rx-example>
+
+<h3><rx-permalink>Input Groups</rx-permalink></h3>
```
Heading should be an `<h4>` to fall under "Intermediate Controls". Please update.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,54 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<rx-example name="forms.advancedcontrols"></rx-example>
+<!-- end: Advanced controls -->
+<!-- end: Advanced text area controls -->
+<h3><rx-permalink>Advanced Text Area</rx-permalink></h3>
```
Heading should be an `<h4>` to fall under "Advanced Controls". Please update.

CITguy commented on src/elements/Forms/docs/examples/forms.intermediatecontrols.js:
```diff
@@ -0,0 +1,36 @@
+// with ngForm to set the appropriate `.$error` value
+// Note: This code is easier to write in Angular 1.3, because
+// you can use `.$validators` instead of `.$parsers`
+angular.module('encore.ui.elements')
```
Please change this module to `demoApp`. This directive does not need to be defined on `encore.ui.elements`, because it's for demo purposes only.

khaziwallis commented:
@CITguy have addressed review comments..

CITguy commented:
The "Advanced Text Area" should have the text area and info side-by-side (as seen at http://rackerlabs.github.io/encore-ui/#/components/rxForm)

### PR Demo
![screen shot 2016-08-09 at 3 24 46 pm](https://cloud.githubusercontent.com/assets/545605/17532269/6fd0488c-5e45-11e6-8650-da486b7449df.png)

### Expected Demo
<img width="1067" alt="screen shot 2016-08-09 at 3 19 24 pm" src="https://cloud.githubusercontent.com/assets/545605/17532139/c78ef84e-5e44-11e6-8095-b360ae6f416c.png">

CITguy commented:
@khaziwallis please address missed comments and Advanced Text Area layout.

CITguy commented:
@khaziwallis: 

Please address these comments:

* https://github.com/rackerlabs/encore-ui/pull/1762#discussion_r73173951
* https://github.com/rackerlabs/encore-ui/pull/1762#discussion_r73177981
* https://github.com/rackerlabs/encore-ui/pull/1762#discussion_r73178043
* https://github.com/rackerlabs/encore-ui/pull/1762#discussion_r73178061

khaziwallis commented:
@CITguy  Addressed review comments for Option Tables,
do i require to spilt Advance Controls into

Radio + Select
checkbox + Text Input
Dropdown + Text Input + Dropdown
Search Box

ssalinas24 commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,119 @@
+
+  </rx-form-section>
+
+  <h4>With <code>controlled-width</code> section attribute</h3>
```
This has different opening and closing tags. Please correct this. @CITguy is this supposed to be an `h4` or `h3`?

CITguy replied:
According to comments, it should be an `<h4>`

CITguy commented:
@khaziwallis: no need to split advanced controls, for now.

CITguy commented on src/elements/Forms/docs/examples/forms.inputgroups.js:
```diff
@@ -0,0 +1,32 @@
+        'Hurt you'
+    ];
+
+    $scope.intermediate = {
```
I think we can simplify this a bit by removing a layer of abstraction.

```javascript
$scope.favoriteBeatle = 'all';
$scope.settings = {
    first: true,
    second: false,
    third: true,
    fourth: false
};
```

CITguy commented on src/elements/Forms/docs/examples/forms.advancedcontrols.js:
```diff
@@ -0,0 +1,7 @@
+angular.module('demoApp')
+.controller('formAdvancedControlsDemoCtrl', function ($scope) {
+    $scope.advanced = {
```
This can be simplified by removing a layer of abstraction:

```javascript
$scope.radChoice = 'default';
$scope.inputEnabled = false;
```

CITguy commented on src/elements/Forms/docs/examples/forms.checkboxOptionTable.js:
```diff
@@ -0,0 +1,39 @@
+        }
+    ];
+
+    $scope.intermediate = {
```
I think we can simplify this a bit by removing a layer of abstraction:

```javascript
$scope.table = {
    checkbox: [true, 'unchecked'], // example with first checkbox automatically checked
};
```

CITguy commented on src/elements/Forms/docs/examples/forms.dropdown.js:
```diff
@@ -0,0 +1,74 @@
+        }
+    ];
+
+    $scope.intermediate = {
```
I think we can simplify this a bit by removing a layer of abstraction:

```javascript
$scope.services = [];
// Select the first volume type by default
$scope.volumeType = _.first($scope.volumeTypes).value;
```

CITguy commented on src/elements/Forms/docs/examples/forms.emptyOptionTable.js:
```diff
@@ -0,0 +1,31 @@
+
+    $scope.optionTableEmptyData = [];
+
+    $scope.intermediate = {
```
We can simplify this a bit by removing a layer of abstraction:

```javascript
$scope.table = {
    empty: [true, 'unchecked']
};
```

CITguy commented on src/elements/Forms/docs/examples/forms.intermediatecontrols.js:
```diff
@@ -0,0 +1,35 @@
+.controller('formIntermediateControlsDemoCtrl', function ($scope) {
+    
+    /* ========== FORM MODELS ========== */
+    $scope.simple = {
```
I think we can simplify this a bit by removing a layer of abstraction:

```javascript
$scope.userEmail = '';
// TODO: use isNameRequired for rxFieldName "required" midway tests
$scope.isNameRequired = true;
$scope.volumeName = '';
```

CITguy commented on src/elements/Forms/docs/examples/forms.intermediatecontrols.js:
```diff
@@ -0,0 +1,35 @@
+angular.module('demoApp')
+.controller('formIntermediateControlsDemoCtrl', function ($scope) {
+    
+    /* ========== FORM MODELS ========== */
```
Comment unnecessary. Please remove.

CITguy commented on src/elements/Forms/docs/examples/forms.intermediatecontrols.js:
```diff
@@ -0,0 +1,35 @@
+// It's used to check that some string contains 'foo', and works
+// with ngForm to set the appropriate `.$error` value
+// Note: This code is easier to write in Angular 1.3, because
+// you can use `.$validators` instead of `.$parsers`
```
Lines 17-18 are no longer necessary with updates suggested below.

CITguy commented on src/elements/Forms/docs/examples/forms.intermediatecontrols.js:
```diff
@@ -0,0 +1,35 @@
+        require: 'ngModel',
+        link: function (scope, elm, attrs, ctrl) {
+            // Put a new validator on the beginning
+            ctrl.$parsers.unshift(function (viewValue) {
```
Since we're using ng1.3 now, we have access to `$validators`, so let's make a quick update to this while we're in here.

```javascript
ctrl.$validators.foo = function (modelValue) {
    return _.contains(modelValue, 'foo');
};
```

CITguy commented on src/elements/Forms/docs/examples/forms.radioOptionTable.js:
```diff
@@ -0,0 +1,65 @@
+        return rowId === 'option4_id';
+    };
+
+    $scope.intermediate = {
```
We can simplify this a bit by removing a layer of abstraction.

```javascript
$scope.table = {
    radio: 0
};
```

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
-</ul>
-<rx-example name="forms.validation"></rx-example>
+<!-- start: rxForm -->
+<!-- start: simple controls -->
```
Comments on line 690-691 are unnecessary. Please remove.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+</p>
+<rx-example name="forms.textarea"></rx-example>
+<!-- end: simple controls -->
+<!-- start: Intermediate controls -->
```
Comments on line 705-706 are unnecessary. Please remove.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<rx-example name="forms.textarea"></rx-example>
+<!-- end: simple controls -->
+<!-- start: Intermediate controls -->
+<h3><rx-permalink>Intermediate Controls</rx-permalink></h3>
```
⬜  Please ensure that there is a blank line above this heading.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<h4><rx-permalink>Input Groups</rx-permalink></h4>
+<rx-example name="forms.inputgroups"></rx-example>
+
+<!-- start: option tables -->
```
Comment is unnecessary. Please remove.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+
+<!-- start: option tables -->
+<h3><rx-permalink>Option Tables</rx-permalink></h3>
+<h4><rx-permalink>Radio Option Tables</rx-permalink></h4>
```
Please add a new line above this heading.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<h3><rx-permalink>Option Tables</rx-permalink></h3>
+<h4><rx-permalink>Radio Option Tables</rx-permalink></h4>
+<rx-example name="forms.radioOptionTable"></rx-example>
+<h4><rx-permalink>Checkbox Option Tables</rx-permalink></h4>
```
Please add a new line above this heading.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<rx-example name="forms.radioOptionTable"></rx-example>
+<h4><rx-permalink>Checkbox Option Tables</rx-permalink></h4>
+<rx-example name="forms.checkboxOptionTable"></rx-example>
+<h4><rx-permalink>Empty Option Tables</rx-permalink></h4>
```
Please add a new line above this heading.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<rx-example name="forms.emptyOptionTable"></rx-example>
+<!-- start: option tables -->
+<!-- end: Intermediate controls -->
+<!-- start: Advanced controls -->
```
Comments on line 730-732 are unnecessary. Please remove.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<!-- start: option tables -->
+<!-- end: Intermediate controls -->
+<!-- start: Advanced controls -->
+<h4><rx-permalink>Advanced Controls</rx-permalink></h4>
```
Please add a new line above this heading.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<h4><rx-permalink>Advanced Controls</rx-permalink></h4>
+<rx-example name="forms.advancedcontrols"></rx-example>
+<!-- end: Advanced controls -->
+<!-- end: Advanced text area controls -->
```
Comments on lines 735-736 are unnecessary. Please remove.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<h3><rx-permalink>Advanced Text Area</rx-permalink></h3>
+<rx-example name="forms.advancedtextarea"></rx-example>
+<!-- end: Advanced text area controls -->
+<!-- start: Form Actions -->
```
Comments on line 739-740 are unnecessary. Please remove.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<rx-example name="forms.advancedtextarea"></rx-example>
+<!-- end: Advanced text area controls -->
+<!-- start: Form Actions -->
+<h3><rx-permalink>Form Actions</rx-permalink></h3>
```
⬜  please ensure that there is a blank line above this heading

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,61 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<h3><rx-permalink>Form Actions</rx-permalink></h3>
+<rx-example name="forms.formactions"></rx-example> 
+<!-- end: Form Actions -->
+<!-- end: rxForm -->
```
Comments on line 743-744 are unnecessary. Please remove.

CITguy commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,119 @@
+            ng-model="demoSearch"
+            rx-placeholder="'Filter by...'"></rx-search-box>
+        </rx-input>
+        <rx-help-text>
```
This should become an `<h5>` and moved above the `<rx-field-name>`.

CITguy commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,119 @@
+            </select>
+          </rx-suffix>
+        </rx-input>
+        <rx-help-text>Dropdown + Text Input + Dropdown</rx-help-text>
```
This should become an `<h5>` and moved above the `<rx-field-name>`.

CITguy commented:
@khaziwallis: more comments added. We're getting close, though. 😄

khaziwallis commented:
@CITguy have made changes as per review.

CITguy commented:
@khaziwallis: Thanks, I'll take a look.

CITguy commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,119 @@
+
+    <rx-field>
+      <h5>
+          Using <code>rx-search-box</code>.
```
For consistency, please put `<h5>` and its contents on one line, like above.

CITguy commented on src/elements/Forms/docs/examples/forms.advancedtextarea.html:
```diff
@@ -0,0 +1,44 @@
+        </p>
+        <p>
+          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed tempus id ligula sit amet 
+          <br>
```
`<br>` tags in this paragraph are unnecessary. Please remove.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,54 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+
+<h4><rx-permalink>Drop-Down Selection</rx-permalink></h4>
+<p>
+  For drop-down selections, two options are available: a drop-down that only permits a single selection, or a drop-down that allows the user to make multiple selections.
```
Per the [HTML Styleguide](https://github.com/rackerlabs/encore-ui/wiki/HTML-Styleguide#html-max-line-length), please break content into multiple lines so that it does not exceed 120 characters wide.

CITguy commented:
It appears that the missing data issue has come back in `Layout > Page Examples > Layout 3: Form Page`. Please correct.

![screen shot 2016-08-12 at 10 36 22 am](https://cloud.githubusercontent.com/assets/545605/17627935/b253ec10-6078-11e6-999e-60b813853e1a.png)

It might be beneficial to create it's own controller file and put it in the `demo/scripts/` directory. That way the `src/` demo content won't interfere with the `demo/` content.

CITguy commented on demo/templates/layout/page/form-page.html:
```diff
@@ -50,8 +50,7 @@
         <tab heading="Tab 1">
-          <!-- Using the controller from the rxForm docs .js file -->
-          <div ng-controller="rxFormDemoCtrl">
+          <div ng-controller="layoutRadioOptionTableCtrl">
```
👍

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,56 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<h4><rx-permalink>Empty Option Tables</rx-permalink></h4>
+<rx-example name="forms.emptyOptionTable"></rx-example>
+
+<h4><rx-permalink>Advanced Controls</rx-permalink></h4>
```
This should be an `<h3>`. This section is not part of "Option Tables" (previous h3). Please update.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,56 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<h4><rx-permalink>Advanced Controls</rx-permalink></h4>
+<rx-example name="forms.advancedcontrols"></rx-example>
+
+<h3><rx-permalink>Advanced Text Area</rx-permalink></h3>
```
This should be an `<h4>` to fall within the "Advanced Controls" section. Please update.

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,56 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<rx-example name="forms.advancedcontrols"></rx-example>
+
+<h3><rx-permalink>Advanced Text Area</rx-permalink></h3>
+<rx-example name="forms.advancedtextarea"></rx-example>
```
Per [this feedback](https://github.com/rackerlabs/encore-ui/pull/1762#discussion_r73352534). Please add a paragraph as follows:

```html
<p>
  Text areas can utilize multiple features simultaneously. In the following 
  example, the text area is a required field, has a character counter 
  applied, and shows help text.
</p>
```

CITguy commented on src/elements/Forms/docs/forms.docs.html:
```diff
@@ -687,26 +687,56 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
+<rx-example name="forms.advancedtextarea"></rx-example>
 
+<h3><rx-permalink>Form Actions</rx-permalink></h3>
+<rx-example name="forms.formactions"></rx-example> 
```
Per [this feedback](https://github.com/rackerlabs/encore-ui/pull/1762#discussion_r73177537), please add the following paragraph (mind the line length).

```html
<p>
  At the end of each form, the user should be presented with a cancel button and a submit button. The 
  submit button should appear on the left, the cancel button to the right. Submit button text should be as 
  specific as possible to reinforce for the user what change they are about to make. For example, "Update 
  Server" is better than a general "Submit".
</p>
```

CITguy commented on src/elements/Forms/docs/examples/forms.dropdown.html:
```diff
@@ -0,0 +1,32 @@
+      <rx-field-name>Service Options:</rx-field-name>
+      <rx-field-content>
+        <rx-input>
+          <rx-multi-select ng-model="emptyServices" id="mselServices">
```
The `ng-model` value should be `services`, as it was when it was defined as `intermediate.services`.

CITguy commented on src/elements/Forms/docs/examples/forms.dropdown.html:
```diff
@@ -0,0 +1,32 @@
+            <rx-select-option ng-repeat="service in services" value="{{service.value}}">{{service.label}}</rx-select-option>
+          </rx-multi-select>
+        </rx-input>
+        <rx-help-text>Bound Value: {{intermediate.services}}</rx-help-text>
```
Bound value should match `ng-model` value above. Please update.

CITguy commented on src/elements/Forms/docs/examples/forms.dropdown.js:
```diff
@@ -0,0 +1,72 @@
+    ];
+    // select the first type by default
+    $scope.volumeType = _.first($scope.volumeTypes).value;
+    $scope.emptyServices = [];
```
This should be `$scope.services = [];`. It is a model value representing saved data, not input data.

khaziwallis replied:
@CITguy i will keep this as selectedServices as 4scope.service is already there. which contains list of services.

CITguy replied:
Good catch. That works. 👍

CITguy commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,117 @@
+            id="radDefault"
+            name="radAdvanced"
+            ng-model="radChoice"
+            value="default" />
```
Please correct indentation for these attributes.

CITguy replied:
The attributes should be indented one level from the start of `<input>`
```html
<input rx-radio
  id="radDefault"
  name="radAdvanced"
  ng-model="radChoice"
  value="default" />
```

CITguy commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,117 @@
+                 id="radCustom"
+                 name="radAdvanced"
+                 ng-model="radChoice"
+                 value="custom" />
```
Please correct indentation of attributes similar to previous comment.

CITguy commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,117 @@
+          </rx-prefix>
+          <input type="text"
+                 ng-disabled="!inputEnabled"
+                 placeholder="Enter Text" />
```
Please correct indentation of attributes similar to previous comment.

CITguy commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,117 @@
+    To prevent fields from taking up too much horizontal space, use the <code>controlled-width</code>
+    <br>
+    attribute on <code>rx-form-section</code>. The fields will still stack if horizontal space is
+    <br>
```
`<br>` tags are unnecessary. Please remove.

CITguy commented on src/elements/Forms/docs/examples/forms.advancedcontrols.html:
```diff
@@ -0,0 +1,117 @@
+  </rx-help-text>
+
+
+
```
Please reduce these three blank lines to one blank line.

CITguy commented on src/elements/Forms/docs/examples/forms.checkboxOptionTable.html:
```diff
@@ -0,0 +1,23 @@
+              field-id="optionCheckboxTable"
+              data="optionTableCheckboxData"
+              required="true"
+              ></rx-option-table>
```
Please correct indentation of attributes. They should be 2 spaces in from the start of `<rx-option-table>`.

CITguy commented on src/elements/Forms/docs/examples/forms.emptyOptionTable.html:
```diff
@@ -0,0 +1,20 @@
+              field-id="optionCheckboxTable"
+              data="optionTableEmptyData"
+              empty-message="You don't have any data!"
+              ></rx-option-table>
```
Please correct indentation of attributes. They should be 2 spaces in from the start of `<rx-option-table>`.

CITguy commented on src/elements/Forms/docs/examples/forms.radioOptionTable.html:
```diff
@@ -0,0 +1,22 @@
+              selected="0"
+              class="full-width"
+              disable-fn="disableOption(tableId, fieldId, rowId)"
+              ></rx-option-table>
```
Please correct indentation of attributes. They should be 2 spaces in from the start of `<rx-option-table>`.

CITguy commented:
@khaziwallis: Comments added. I believe I've captured feedback that may have been overlooked in the plethora of comments, too.

CITguy commented:
LGTM Thank you for your patience.

CITguy commented:
@khaziwallis: please squash to one commit

khaziwallis commented:
@CITguy have done squash and amend.

parlarjb commented:
LGTM

ElementE commented:
Design LGTM
